dbms/src/Storages/StorageReplicatedMergeTree.h

```diff
@@ -246,22 +249,27 @@ class StorageReplicatedMergeTree : public ext::shared_ptr_helper<StorageReplicat
 
     /// Threads.
 
-    /// A thread that keeps track of the updates in the logs of all replicas and loads them into the queue.
-    std::thread queue_updating_thread;
-    zkutil::EventPtr queue_updating_event = std::make_shared<Poco::Event>();
+    /// A task that keeps track of the updates in the logs of all replicas and loads them into the queue.
+    bool queue_update_in_progress = false;
+    BackgroundSchedulePool::TaskHandle queue_updating_task_handle;
 
     /// A task that performs actions from the queue.
     BackgroundProcessingPool::TaskHandle queue_task_handle;
 
-    /// A thread that selects parts to merge.
-    std::thread merge_selecting_thread;
-    Poco::Event merge_selecting_event;
+    /// A task that selects parts to merge.
+    BackgroundSchedulePool::TaskHandle merge_selecting_handle;
+    bool merge_sel_deduplicate;
+    bool merge_sel_need_pull;
+    std::function<bool(const MergeTreeData::DataPartPtr &, const MergeTreeData::DataPartPtr &)> merge_sel_uncached_merging_predicate;
+    std::function<std::pair<String, String>(const MergeTreeData::DataPartPtr &, const MergeTreeData::DataPartPtr &)> merge_sel_merging_predicate_args_to_key;
+    std::chrono::steady_clock::time_point merge_sel_now;
+    std::unique_ptr<CachedMergingPredicate<std::pair<std::string, std::string>> > merge_sel_cached_merging_predicate;
+    std::function<bool(const MergeTreeData::DataPartPtr &, const MergeTreeData::DataPartPtr &)> merge_sel_can_merge;
```

Comment:
The fact that we need a lot of fields here with the common prefix suggests that we should encapsulate the context of the merge selecting task in its own class.

Reply:
@ztlpn looking to the code seems not so trivial. merge_sel_uncached_merging_predicate and merge_sel_can_merge are functions that have in their implementations lot of members from StorageReplicatedMergeTree. 

How you prefer to deal with this ? I want to avoid to pass to the new context class a reference to StorageReplicatedMergeTree  and to make them friend calsses (as I saw into lot of places in the code) because I have to violate the encapsulation. And code won't be better than now..

Reply:
I think even a separate class that is a friend of StorageReplicatedMergeTree is better in terms of encapsulation. Yes, it can access private members of StorageReplicatedMergeTree, but StorageReplicatedMergeTree can't access its private members! As opposed to the current solution where everybody can access anything.

Reply:
Should I add the new class `StorageReplicatedMergeTreeContext` into same header StorageReplicatedMergeTree.h  or do you want a separate file ?.

Reply:
I think we should add ReplicatedMergeTreeMergeSelectingThread in the separate file in the Storages/MergeTree dir (just like e.g. ReplicatedMergeTreeCleanupThread)

Reply:
Well in some sense these tasks are still sequential threads of execution, just not represented as OS threads... And I agree that renaming it now would be painful.

Reply:
Ok I will make some time this days to fix this (from what I see last remaining task). I will call the new class ReplicatedMergeTreeMergeSelectingThread and place it in Storages/MergeTree dir.

I also need to bring the branch up to date again with master. I see an ugly conflict there.. need to see how I have to fix.

Reply:
Hello, 

In commit f811da7ed4ef1fa140415bd9aa5d009334170237 you can find the version of the fix where all merge_sel_* are moved into ReplicatedMergeTreeMergeSelectingThread

What I didn't done yet is to move this class into it's own file. Here I have some problems that I can fix in many ways 
and not sure how you prefer.

Problem is that: canMergePartsAccordingToZooKeeperInfo it's used also into line 2456 and this function is locally only in this file.

What I can do is to move canMergePartsAccordingToZooKeeperInfo, partsWillNotBeMergedOrDisabled, CachedMergingPredicate into the new file and 
	1) for first function I can make it member of ReplicatedMergeTreeMergeSelectingThread so can be used from StorageReplicatedMergeTree as well
	2) create in ReplicatedMergeTreeMergeSelectingThread another version for can_merge similar with the one from StorageReplicatedMergeTree line 2456	
		
Please review this commit and also let me know how to proceed. One of the above options or another idea		

Silviu

Reply:
@ztlpn any input on this ? want to finish it and then to merge with master as time lot of conflicts are already accumulating